packages/apollo-link-http/src/__tests__/httpLink.ts

```diff
@@ -215,25 +213,221 @@ describe('HttpLink', () => {
     }, 50);
   });
 
-  it('should add headers from the context', done => {
-    const fetch = createApolloFetch({
-      customFetch: (request, options) =>
-        new Promise((resolve, reject) => {
-          expect(options.headers['test']).toBeDefined();
-          expect(options.headers.test).toEqual(context.headers.test);
-          done();
-        }),
+  it('adds headers to the request from the context', done => {
+    const variables = { params: 'stub' };
+    const middleware = new ApolloLink((operation, forward) => {
+      operation.setContext({
+        headers: { authorization: '1234' },
+      });
+      return forward(operation);
     });
-    const link = new HttpLink({ fetch });
+    const link = middleware.concat(createHttpLink({ uri: 'data' }));
+
+    execute(link, { query: sampleQuery, variables }).subscribe(result => {
+      const headers = fetchMock.lastCall()[1].headers;
+      expect(headers.authorization).toBe('1234');
+      expect(headers['content-type']).toBe('application/json');
+      expect(headers.accept).toBe('*/*');
+      done();
+    });
+  });
+  it('adds headers to the request from the context on an operation', done => {
+    const variables = { params: 'stub' };
+    const link = createHttpLink({ uri: 'data' });
 
     const context = {
-      headers: {
-        test: 'header',
+      headers: { authorization: '1234' },
+    };
+    execute(link, {
+      query: sampleQuery,
+      variables,
+      context,
+    }).subscribe(result => {
+      const headers = fetchMock.lastCall()[1].headers;
+      expect(headers.authorization).toBe('1234');
+      expect(headers['content-type']).toBe('application/json');
+      expect(headers.accept).toBe('*/*');
+      done();
+    });
+  });
+  it('adds creds to the request from the context', done => {
+    const variables = { params: 'stub' };
+    const middleware = new ApolloLink((operation, forward) => {
+      operation.setContext({
+        credentials: 'same-team-yo',
+      });
+      return forward(operation);
+    });
+    const link = middleware.concat(createHttpLink({ uri: 'data' }));
+
+    execute(link, { query: sampleQuery, variables }).subscribe(result => {
+      const creds = fetchMock.lastCall()[1].credentials;
+      expect(creds).toBe('same-team-yo');
+      done();
+    });
+  });
+  it('adds fetcherOptions to the request from the context', done => {
+    const variables = { params: 'stub' };
+    const middleware = new ApolloLink((operation, forward) => {
+      operation.setContext({
+        fetcherOptions: {
+          signal: 'foo',
+        },
+      });
+      return forward(operation);
+    });
+    const link = middleware.concat(createHttpLink({ uri: 'data' }));
+
+    execute(link, { query: sampleQuery, variables }).subscribe(result => {
+      const signal = fetchMock.lastCall()[1].signal;
+      expect(signal).toBe('foo');
+      done();
+    });
+  });
+});
+
+describe('dev warnings', () => {
```

Comment:
WOAH thank you for writing all these tests, next time I will tackle them :) they look great